Frontend/Testing.md

```diff
@@ -0,0 +1,336 @@
+  expect(isOddNumber(35))
+    .toEqual(true);
+});
+```
+
+How to use tools in each stack is something each stack should cover.
+
+### TDD
+
+We need to note that we will simplify TDD concept since it cannot be explained in single page.
+
+So what is TDD?
+
+TDD or test driven development is a technique of writing tests before code in a way that you write
+simple failing test and as smallest code possible to pass that test.
+After that you write another simple failing test and another smallest code to pass this second one.
+You continue to do that until you are finished with code and refactor in the middle.
+
+Sounds confusing, but it is not.
+Let's see an example.
+
+We will not write code, just explain what is happening.
+Let's say we want to develop calculator function that allows addition, subtraction, multiplication and
+division.
+calculator function should take 3 arguments: two numbers and operation type.
+And it needs to be bulletproof and check for all invalid usage.
+
+We will use TDD and start with simple failing test:
+
+- calculator should throw error if called empty
+
+Why is this test failing?
+Since we didn't write code to fulfill that test statement yet.
+We write some code to throw error if calculator called empty.
+Test passes.
+
+Next simple failing test will be:
+
+- calculator should throw error if called with less or more than 3 arguments
+
+We could also have negation like: calculator should not throw error if there is exactly 3 arguments.
+Try to avoid negations since they can add confusion to code.
+Using clean code principles here.
+And negativity is not good.
+
+Back to test statement.
+We write code to throw error if there is 0, 1, 2, 4 or more arguments.
+Test passes.
+
+We see opportunity to refactor both in code and test.
+There is check for empty and for exactly three arguments.
+We can leave them both, or remove empty check since it is covered by this second test.
+
+Let's remove empty check to clean code.
+Next we need to check if first and second arguments are numbers, and last one string.
+
+- calculator should throw error if invalid types passed for arguments
+
+We write code to check for types.
+Test passes.
+
+Then
+
+- calculator should throw error if operation type is invalid
+
+We test that string is correct value which can be 'ADD', 'SUBTRACT' and so on - whatever our
+interface will provide.
+Code will be added and test passes.
+
+Then there is that HUGE check if dividing with 0.
+
+- calculator should not have 0 for 2nd number with 'DIV' operation type
+
+Code for that and test passes.
+
+You might be yelling 'Where is calculation implementation!'.
+We go step by step and here we first assure that usage is correct, and then functionality should be
+added.
+
+Can you see how these checks are describing usage similar to your requirements?
+This is very important.
+Imagine someone goes to your tests first.
+He can know what your code does, if we assume that tests pass, without going to source.
```

Comment:
```suggestion
They can know what your code does, if we assume that tests pass, without going to source.
```
Better to use they instead of he/she since it's more inclusive.